### PR TITLE
Task/accidents node aggregation

### PR DESCRIPTION
**Pull request by gry-mar**
Moduł do agregacji wypadków z analizą i użyciem w dedykowanym notebooku
- cechy wierzchołków z OSMNx
- agregacja na podstawie obszaru aż nie zagreguje się wszystkich wypadków

### CODE REVIEW

File: src/graph/create_osmnx_graph.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+        pyg_graph = from_networkx(self.graph_nx)
+        # x and y are node attrs to assign edge attributes use 'pyg_graph.edge_attr'
+        pyg_graph.y = torch.tensor(self.gdf_nodes['accidents_count'].values, dtype=torch.long)
+        features = self._get_node_features()
+        pyg_graph.x = torch.tensor(features.values, dtype=torch.float32)
+        self.graph_data = pyg_graph
+        return pyg_graph
+        
+    
+    # def graph_visualization(self):
```
Comment (JuliaJ194) [inline src/graph/create_osmnx_graph.py:132]: to ma zostać?

File: src/graph/create_osmnx_graph.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+        
+    #     if self.weights=='centroid' or self.weights=='shortest_path':
+    #         width = list(nx.get_edge_attributes(self.graph_nx, 'weight').values())
+    #         nx.draw(self.graph_nx, node_size=30, node_color='plum', width=width)
+    #         plt.show()
+    #     else:
+    #         nx.draw(self.graph_nx, node_size=30, node_color='plum', width=width)
+    #         plt.show()
+    
+    def show_statistics(self):
```
Comment (JuliaJ194) [inline src/graph/create_osmnx_graph.py:144]: moze by to zwracać jako np słownik? żeby później moc uwyprintowac w notebooku albo logować w wandb

File: src/graph/create_osmnx_graph.py (hunk 1 above)
Comment (JuliaJ194) [reply src/graph/create_osmnx_graph.py:132]: też można zwracać jako fig

File: src/graph/create_osmnx_graph.py (hunk 2 above)
Comment (gry-mar) [reply src/graph/create_osmnx_graph.py:144]: można, zmienię to

File: src/graph/create_osmnx_graph.py (hunk 1 above)
Comment (gry-mar) [reply src/graph/create_osmnx_graph.py:132]: właśnie nie wiedziałam co zrobić z tą wizualizacją tak szczerze, bo sam plot trochę długo leci. chyba lepiej byłoby zwrócić po prostu graf NX i w utilsach do wizualizacji już zrobić metody wizualizacji, łatwiej będzie później decydować w jaki sposób wgl to prezentować. co myślisz?

File: src/graph/create_osmnx_graph.py (hunk 2 above)
Comment (JuliaJ194) [reply src/graph/create_osmnx_graph.py:144]: mogę zmienić, bo do wandb tez mi sie przyda

Comment (gry-mar) [reply src/graph/create_osmnx_graph.py:144]: okej, to ja to zostawiam w takim razie

File: src/graph/create_graph.py [hunk 3]
```diff
@@ -113,14 +113,11 @@ def show_statistics(self):
             else data.num_nodes * (data.num_nodes - 1) // 2
         )
         edges = data.num_edges if data.is_directed() else data.num_edges // 2
-        print("Statistics:")
-        print("-" * 50)
-        print("Nodes: ", data.num_nodes)
-        print("Edges: ", edges)
-        print(
```
Comment (JuliaJ194) [inline src/graph/create_graph.py:120]: dupa

File: src/graph/create_osmnx_graph.py [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+import torch
+import pandas as pd
+import osmnx as ox
+import math
+from sklearn.feature_extraction.text import CountVectorizer
+import numpy as np
+from shapely.geometry import box
+from scipy.spatial.distance import cdist
+
+class OSMnxGraph:
```
Comment (StaszekM) [inline src/graph/create_osmnx_graph.py:13]: udokumentuj w tej klasie te funkcje, opisz pokrótce jak dzialają, ciężko zrozumieć co się tu dzieje a to są nietrywialne algorytmy i każdy musi bez wczytywania się w kod rozumieć, to przyspieszy pracę i używanie

sam w PR tutaj proponuję format docstringa wg NumPy: https://github.com/StaszekM/gradient/pull/26

Dlaczego NumPy? Bo moim zdaniem wygląda bardziej zrozumiale niż Google docstring czy inne formaty, NumPy i inne scientific biblioteki lubią to, a do tego współpracuje ze Sphinx więc gdy będziemy kiedyś chcieli zrobić profesjonalną stronkę do dokumentacji taką jaką mają Pandas czy ScikitLearn, to będzie łatwe

Jest gotowa wtyczka do VScode która generuje docstringi: https://marketplace.visualstudio.com/items?itemName=njpwerner.autodocstring

Materiały dotyczącye NumPy docstring:
https://numpydoc.readthedocs.io/en/latest/format.html
https://www.datacamp.com/tutorial/docstrings-python

File: src/graph/create_osmnx_graph.py [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+            self.gdf_edges['accidents_count'] = 0
+        else:
+            raise ValueError("Invalid aggregation_type. Choose either 'node' or 'edge'.")
+    
+        square_edge_length = self.start_acc_distance
+
+        # Iterate through accidents
+        for _, accident in self.gdf_accidents.iterrows():
+            # Create square around accident point
+            accident_point = accident.geometry
```
Comment (StaszekM) [inline src/graph/create_osmnx_graph.py:81]: myślę że tworzenie boxa i szukanie elementów wewnątrz box niepotrzebnie się duplikuje. Spróbuj może zrobić to w jednej pętli while z breakiem? myślę że będzie spoko

zwróć uwagę że pierwsze odpalenie tej logiki przed while niewiele się różni od kolejnych odpaleń wewnątrz while, może tylko że wzrasta width boxa

File: src/graph/create_osmnx_graph.py [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+        
+        self.gdf_accidents = gdf_accidents
+        self.gdf_nodes = gdf_nodes
+        self.gdf_edges = gdf_edges
+        self.start_acc_distance = start_acc_distance
+        self.graph_nx = None
+        self.graph_data = None
+    
+
+    def _get_lat_lon_distance(self, lat, lon, meters):
```
Comment (StaszekM) [inline src/graph/create_osmnx_graph.py:28]: z reguły kolejność deklaracji jest taka:

1. konstruktor
2. metody publiczne
3. metody prywatne z _

Prywatne nie zawsze interesują developera i dlatego lądują na dole

File: src/graph/create_osmnx_graph.py [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+    def _find_nearest_node(accident_point, edges):
+        distances = {}
+        for id, edge in zip(edges.index, edges):
+            distance = edge.distance(accident_point)
+            distances[id] = distance
+        sorted_dict = dict(sorted(distances.items(), key=lambda item: item[1]))
+        first_element = next(iter(sorted_dict.items()))
+        return first_element[0]
+    
+    def _aggregate_accidents(self, aggregation_type:str):
```
Comment (StaszekM) [inline src/graph/create_osmnx_graph.py:65]: jak oceniasz szybkość tej funkcji? pytam bo widzę sporo for, while, a także niskopoziomowe funkcje na shapely

File: src/graph/create_osmnx_graph.py [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+    def _get_edge_attrs(self):
+        """
+        Edge features out of osmnx.
+        """
+        result_df = self.gdf_edges.drop(['geometry', 'u', 'v', 'accidents_count', 'ref', 'width'], axis=1)
+        result_df.fillna('unknown', inplace=True)
+        return result_df
+        
+        
+    def create_graph(self, aggregation_type):
```
Comment (StaszekM) [inline src/graph/create_osmnx_graph.py:149]: aggregation type to chyba będzie union jakby co

File: src/graph/create_osmnx_graph.py [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+        new_longitude_W = lon - (meters * m) / math.cos(lat * (pi / 180))
+        
+        return new_latitude_N, new_latitude_S, new_longitude_W, new_longitude_E
+    
+    
+    def _find_nearest_node(self, accident_point, nodes):
+        accident_point = np.array(accident_point.xy).T[0]
+        distances = {}
+        for id, node in zip(nodes.index, nodes):
+            distance = cdist([accident_point], [node])
```
Comment (StaszekM) [inline src/graph/create_osmnx_graph.py:50]: w tym cdist da się może wyliczyć wszystkie odległości parami naraz? bo tutaj robisz to element po elemencie, a w dokumentacji widzę że można robić to na całych arrayach

File: src/graph/create_osmnx_graph.py [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+        df_count = pd.DataFrame(X.toarray(), columns=vectorizer.get_feature_names_out())
+
+        # Concatenate with original DataFrame if necessary
+        result_df = pd.concat([features, df_count], axis=1)
+        result_df.fillna(0, inplace=True)
+        result_df.drop(['highway', 'unknown'], axis=1, inplace=True)
+        return result_df
+    
+    
+    def _get_edge_attrs(self):
```
Comment (StaszekM) [inline src/graph/create_osmnx_graph.py:140]: może wybierz jedną z opcji nazewnicwa, tj attrs or features?

**Review by StaszekM — CHANGES_REQUESTED** [review]

File: src/graph/create_osmnx_graph.py (hunk 8 above)
Comment (gry-mar) [reply src/graph/create_osmnx_graph.py:149]: nie rozumiem

File: src/graph/create_osmnx_graph.py (hunk 5 above)
Comment (gry-mar) [reply src/graph/create_osmnx_graph.py:81]: To jak proponujesz to rozwiązać? bo szczerze mówiąc nie mam pomysłu

**Comment by gry-mar** [comment]
@StaszekM możesz sprawdzić czy ok

File: notebooks/osmnx_graph_train_test.ipynb
Comment (StaszekM) [inline notebooks/osmnx_graph_train_test.ipynb:1]: W tym notebooku trzeba dopisać w markdown poszczególne fazy i co się dzieje, bo czytanie kodu ujawnia tylko co jest robione, a nie ujawnia *po co* - wiem że w niektórych miejscach to zrobiłaś, ale nie we wszystkich, np przygotowywanie modelu, trening, ewaluacja

dodatkowo usuń outputy z pytorch lightninga bo to są setki linijek i zostają w gicie

Comment (StaszekM) [inline notebooks/osmnx_graph_train_test.ipynb:1]: Widzę kilka razy kawałki kodu z nazwami zmiennych typu X2, gdf2 - zmień proszę nazwy na bardziej deskryptywne

importy wewnątrz jakichś komórek niżej przerzuć do góry żeby trzymać się naszych ustaleń

Comment (StaszekM) [reply notebooks/osmnx_graph_train_test.ipynb:1]: usuń wszystkie komórki które pokazują jakieś outputy, które nie są niezbędne - rozumiem że sobie używałaś tych komórek żeby sprawdzać różne rzeczy po drodze, ale tu są niepotrzebne i utrudniają śledzenie wątku

File: src/baseline_models/GCN/supervised_node_classification.py [hunk 11]
```diff
@@ -107,7 +107,7 @@ def training_step(
         self.log("step", self.trainer.current_epoch)
         self.log("train/loss", loss.item(), on_epoch=True, on_step=False)
         # self.log("train/auc", auc.item(), on_epoch=True, on_step=False)
-
+        print(f"loss: {loss.item()}")
```
Comment (StaszekM) [inline src/baseline_models/GCN/supervised_node_classification.py:110]: po co ten print? linijkę wyżej loguje się do loggera

File: src/graph/create_osmnx_graph.py [hunk 12]
```diff
@@ -0,0 +1,276 @@
...
+        self.graph_data = None
+
+
+    
+    def get_node_attrs(self):
+        """Method to extract node features from OSMNx nodes GeoDataFrame. In that case retrieves hghway types and street_count
+        for each node. 
+
+        Returns:
+            pd.DataFrame: cleaned dataframe that contains node features
```
Comment (StaszekM) [inline src/graph/create_osmnx_graph.py:35]: na czym polega słowo 'cleaned'? jaki 'cleaning' jest robiony?

File: src/graph/create_osmnx_graph.py [hunk 13]
```diff
@@ -0,0 +1,276 @@
...
+        self.gdf_nodes = result_df
+        return result_df
+    
+
+
+    def get_edge_attrs(self):
+        """Method that gets edge attributes.
+
+        Returns:
+            pd.DataFrame: cleaned DataFrame that contains edge features
```
Comment (StaszekM) [inline src/graph/create_osmnx_graph.py:59]: jw. cleaning ?

File: src/graph/create_osmnx_graph.py [hunk 14]
```diff
@@ -0,0 +1,276 @@
...
+        result_df = pd.merge(attrs, df_count, left_index=True, right_index=True)
+        result_df.fillna(0, inplace=True)
+        result_df.drop(['highway', 'unknown'], axis=1, inplace=True)
+        self.gdf_nodes = result_df
+        return result_df
+    
+
+
+    def get_edge_attrs(self):
+        """Method that gets edge attributes.
```
Comment (StaszekM) [inline src/graph/create_osmnx_graph.py:56]: tu zdecydowanie musisz opisać więcej rzeczy które nie wynikają prosto z nazwy funkcji, to że to jest geter na edge attrs to już wiemy xdd

File: src/graph/create_osmnx_graph.py [hunk 15]
```diff
@@ -0,0 +1,276 @@
...
+            coord2 (Point): second coord
+
+        Returns:
+            distance: distance between coordinates
+        """
+        return cdist([coord1], [coord2])[0, 0]
+    
+    
+    def _find_nearest_node(self, accident_point, nodes):
+        """Finds nearest node
```
Comment (StaszekM) [inline src/graph/create_osmnx_graph.py:181]: ...nearest node from given accident point

File: src/graph/create_osmnx_graph.py [hunk 16]
```diff
@@ -0,0 +1,276 @@
...
+            osmid
+        """
+        accident_point = np.array(accident_point.xy).T[0]
+        accident_series = pd.Series([accident_point] * len(nodes), index=nodes.index)
+        sorted_nodes = nodes.combine(accident_series, self._calculate_distance).sort_values()
+        first_dist_osmid = sorted_nodes.index[0]
+        return first_dist_osmid
+    
+    def _find_nearest_edge(self, accident_point, edges):
+        """Finds nearest edge
```
Comment (StaszekM) [inline src/graph/create_osmnx_graph.py:197]: ...nearest edge from given accident point

File: src/graph/create_osmnx_graph.py (hunk 5 above)
Comment (StaszekM) [reply src/graph/create_osmnx_graph.py:229]: to zostawmy jak jest

File: src/graph/create_osmnx_graph.py (hunk 14 above)
Comment (StaszekM) [reply src/graph/create_osmnx_graph.py:56]: sory za mansplaining albo jeśli mówię ci coś co już wiesz, ale dobre komentarze tłumaczą tak, by człowiek bez patrzenia w kod wiedział, jak tego użyć, co zostanie zwrócone i co się stanie po drodze

File: .DS_Store
Comment (StaszekM) [inline .DS_Store:1]: co to?

**Review by StaszekM — CHANGES_REQUESTED** [review]

File: notebooks/osmnx_graph_train_test.ipynb
Comment (gry-mar) [reply notebooks/osmnx_graph_train_test.ipynb:1]: spoko, w sumie traktowałam ten notebook bardziej roboczo ale no widzę że to bez sensu

File: src/baseline_models/GCN/supervised_node_classification.py (hunk 11 above)
Comment (gry-mar) [reply src/baseline_models/GCN/supervised_node_classification.py:110]: no żeby sobie testowo zobaczyć ale usunę

File: .DS_Store
Comment (gry-mar) [reply .DS_Store:1]: ah jakieś macowe coś chyba, pierwszy raz w sumie pracowałam na gradiencie na macu i się musiało dodać, dam do .gitignore
